### PR TITLE
Remove ScriptClassInfo from Walker

### PR DESCRIPTION
**Pull request by jdconrad**
This change covers two items:
1. This removes ScriptClassInfo from Walker so that we can separate semantic checking from user tree building entirely. ScriptClassInfo is added to a PainlessSemanticAnalysisPhase and a PainlessSemanticHeaderPhase which both extend from their respective Default classes. These classes allow for the special casing of the execute method where ScriptClassInfo is used to inject the appropriate parameters for the execute method. This allows us to potentially check and use a script across more than one context which is especially relevant for stored scripts.
2. This removes the BootstrapInjectionPhase and the ScriptInjectionPhase instead moving the code more appropriately to the other phases instead.

### CODE REVIEW

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/bwc

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultSemanticAnalysisPhase.java [hunk 1]
```diff
@@ -261,13 +261,6 @@ public void visitFunction(SFunction userFunctionNode, ScriptScope scriptScope) {
         if (methodEscape) {
             functionScope.setCondition(userFunctionNode, MethodEscape.class);
         }
-
-        // TODO: do not specialize for execute
-        // TODO: https://github.com/elastic/elasticsearch/issues/51841
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultSemanticAnalysisPhase.java:266]: Does this close #51841?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 2]
```diff
@@ -176,26 +179,187 @@
...
-        } else {
-            userNode.visit(this, scriptScope);
-            return scriptScope.getDecoration(userNode, IRNodeDecoration.class).getIRNode();
+    /**
+     * This injects additional ir nodes required for resolving the def type at runtime.
+     * This includes injection of ir nodes to add a function to call
+     * {@link DefBootstrap#bootstrap(PainlessLookup, FunctionTable, Lookup, String, MethodType, int, int, Object...)}
+     * to do the runtime resolution.
+     */
+    protected void injectBootstrapMethod(ScriptScope scriptScope) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:207]: Should you mention `injectStaticFields` here?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java [hunk 3]
```diff
@@ -332,11 +388,11 @@ protected static void injectSandboxExceptions(FunctionNode functionNode) {
             memberCallNode.setLocation(internalLocation);
             memberCallNode.setExpressionType(ScriptException.class);
             memberCallNode.setLocalFunction(new LocalFunction(
-                    "convertToScriptException",
-                    ScriptException.class,
-                    Arrays.asList(Throwable.class, Map.class),
-                    true,
-                    false
+                            "convertToScriptException",
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java:391]: remove indent

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java [hunk 4]
```diff
@@ -367,15 +423,15 @@ protected static void injectSandboxExceptions(FunctionNode functionNode) {
...
-                    PainlessExplainError.class,
-                    null,
-                    Collections.emptyList(),
-                    null,
-                    null,
-                    null
+                            PainlessExplainError.class.getMethod(
+                                    "getHeaders",
+                                    PainlessLookup.class),
+                            PainlessExplainError.class,
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java:429]: remove indent

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java [hunk 5]
```diff
@@ -418,11 +474,11 @@ protected static void injectSandboxExceptions(FunctionNode functionNode) {
...
                 memberCallNode.setLocalFunction(new LocalFunction(
-                        "convertToScriptException",
-                        ScriptException.class,
-                        Arrays.asList(Throwable.class, Map.class),
-                        true,
-                        false
+                                "convertToScriptException",
+                                ScriptException.class,
+                                Arrays.asList(Throwable.class, Map.class),
+                                true,
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java:480]: remove indent

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java [hunk 6]
```diff
@@ -452,13 +508,13 @@ protected static void injectSandboxExceptions(FunctionNode functionNode) {
...
-                        null,
-                        null,
-                        null
+                                Collections.class.getMethod("emptyMap"),
+                                Collections.class,
+                                null,
+                                Collections.emptyList(),
+                                null,
+                                null,
+                                null
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java:517]: remove indent

**Review by stu-elastic — APPROVED** [review]
good job despecializing `execute`

**Comment by jdconrad** [comment]
@stu-elastic Thanks for the review! Will address PR comments shortly.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultSemanticAnalysisPhase.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultSemanticAnalysisPhase.java:266]: Yes!

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:207]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java (hunk 3 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java:391]: Done

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java (hunk 4 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java:429]: Done

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java (hunk 5 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java:480]: Done

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java (hunk 6 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java:517]: Done

**Pull request merged**